DllImportGenerator/DllImportGenerator/ManualTypeMarshallingAnalyzer.cs

```diff
@@ -0,0 +1,378 @@
+                }
+
+                ITypeSymbol nativeType = (ITypeSymbol)nativeMarshalerAttributeData.ConstructorArguments[0].Value!;
+
+                if (!nativeType.IsValueType)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustHaveRequiredShapeRule, GetSyntaxReferenceForDiagnostic(nativeType).GetSyntax().GetLocation(), nativeType.ToDisplayString(), type.ToDisplayString()));
+                    return;
+                }
+
+                if (nativeType is not INamedTypeSymbol marshalerType)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustHaveRequiredShapeRule, nativeMarshalerAttributeData.ApplicationSyntaxReference!.GetSyntax().GetLocation(), nativeType.ToDisplayString(), type.ToDisplayString()));
+                    return;
+                }
+
+                bool hasConstructor = false;
+                bool hasStackallocConstructor = false;
+                foreach (var ctor in marshalerType.Constructors)
+                {
+                    if (ctor.IsStatic)
+                    {
+                        continue;
+                    }
+                    if (ctor.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals(type, ctor.Parameters[0].Type))
+                    {
+                        hasConstructor = true;
+                    }
+                    if (ctor.Parameters.Length == 2 && SymbolEqualityComparer.Default.Equals(type, ctor.Parameters[0].Type)
+                                                    && SymbolEqualityComparer.Default.Equals(SpanOfByte, ctor.Parameters[1].Type))
+                    {
+                        hasStackallocConstructor = true;
+                    }
+                }
+
+                bool hasToManaged = marshalerType.GetMembers("ToManaged")
+                                                    .OfType<IMethodSymbol>()
+                                                    .Any(m => m.Parameters.IsEmpty && !m.ReturnsByRef && !m.ReturnsByRefReadonly && SymbolEqualityComparer.Default.Equals(m.ReturnType, type) && !m.IsStatic);
+
+                if (!hasConstructor && !hasStackallocConstructor && !hasToManaged)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustHaveRequiredShapeRule, GetSyntaxReferenceForDiagnostic(marshalerType).GetSyntax().GetLocation(), marshalerType.ToDisplayString(), type.ToDisplayString()));
+                }
+
+                if (validateAllScenarioSupport && hasStackallocConstructor && !hasConstructor)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(StackallocMarshallingShouldSupportAllocatingMarshallingFallbackRule, GetSyntaxReferenceForDiagnostic(marshalerType).GetSyntax().GetLocation(), marshalerType.ToDisplayString()));
+                }
+
+                IPropertySymbol? valueProperty = nativeType.GetMembers("Value").OfType<IPropertySymbol>().FirstOrDefault();
```

Comment:
Add a basic comment for the rule / what we're checking for in the different parts of this method? Just to make it easier to scan through the method quickly.